proposals/0000-complete-set-signatures.rst

```diff
@@ -0,0 +1,483 @@
+match type that satisfies ``IsInfinite`` that a
+user might write.
+
+That is the new feature that I propose. Note the
+specification of the full result type instead of
+just the name of a data type constructor: It's a
+breaking change with the current syntax, but one
+that is trivially flagged by a kinding error.
+
+Proposed Change Specification
+-----------------------------
+
+Syntax
+======
+
+Extend Syntax so that we are able to write the example from the previous section
+
+::
+
+ {-# COMPLETE (:<), Empty :: IsList l => l a #-} -- Context is allowed
+ {-# COMPLETE (:<) :: Stream a #-}               -- TyCon App is allowed
+ {-# COMPLETE (:<), Empty :: String #-}          -- Even type synonyms
+
+Since there is no formal grammar for COMPLETE pragmas, here's how I propose to
+change the happy grammar from
+
+::
+
+ sigdecl
+   : ...
+   | '{-# COMPLETE' con_list opt_tyconsig  '#-}'
+ opt_tyconsig
+   : {- -}
+   | '::' gtycon
+ gtycon
+   : ntgtycon
+   | '(' ')'
+   | '(#' '#)'
+
+to
+
+::
+
+ sigdecl
+   : ...
+   | '{-# COMPLETE' con_list opt_restysig  '#-}'
+ opt_restysig
+   : {- -}
+   | '::' ctype
+ atype
+   : ntgtycon
+   | '(' ')'
+   | '(#' '#)'
+   | ... many more ...
+
+Where a ``ctype`` is a "for-all type", according to GHC's happy parser. It's the
+same grammatical sort that constitutes the RHS of a ``::`` in a type signature
+and since it constrains the result type of the relevant ConLikes, we call
+it a "result type signature", often referring just to the type it carries.
+
+Note that this change accepts strictly more syntax, because
+
+- ``atype`` accepts a superset of ``gtycon``
+- ``atype`` can be derived from ``ctype`` (via ``type``, ``btype``, ``infixtype``, ``ftype``)
+
+Typing
+======
+
+The (type carried by the) result type signature must have kind ``TYPE r``, for
+any runtime-representation ``r``. Any free type variables are implicitly bound
+by top-level ``forall``, just like it happens for function type signatures.
+
+Examples for well-typed result type signatures:
+
+::
+
+ [a]
+ IsInfinite l => l a
+ forall l a. IsInfinite l => l a  -- equivalent to the former
+ String
+ (a ~ Int, Semigroup a) => a
+ MPTC a b => a
+ Int#
+
+Examples for invalid result type signatures:
+
+::
+
+ Stream
+ (->)
+ "Symbol"
+ Eq Int
+
+Note that after type-checking
+
+- We *accept* some of the previously well-typed syntax, like ``Int`` and other
+  nullary data type constructors.
+- We *reject* some of the previously well-typed syntax, like ``Stream`` and other
+  non-nullary data type constructors.
+- We *accept* new syntax, like ``IsInfinite l => l a``.
+
+Pattern match checking
+======================
+
+A COMPLETE set with a result type signature ``{-# COMPLETE cls :: sig_ty
+#-}`` is to be treated the same as one without, with one exception: When
+the COMPLETE set is supposedly *covered* by a set of patterns in a pattern
+match, we
+
+1. Take the result type of the pattern match, ``ty``.
```

Comment:
This part is very hazy.   Consider
```
f :: Eq a => Maybe (T a) -> blah
f (Just (K1 x)) = ...
f (Just (K2 p q)) = ...
...
```
Here I think you want to say that the "result type of the pattern match" is `forall a. Eq a => T a`, even though no term has that type.    Yes, the `K x` has type `T a`, but adding the constraints and forall is an informal thing.

To see the informality, note the the forall'd variable can be existential.
```
data X where { MkX :: Eq x => T x ->X }   -- x is existential

f :: X -> blah
f (MkX (K1 x)) = ...
f (MkX (K2 p q)) = ...
```
Here also I think you want to say that the "result type of the pattern match" is `forall x. Eq x => T x`, even though `x` is "really" existential.  

To to get the "result type of the pattern match" I think we have to take
* The result type of the pattern match itself
* Add all the in-scope "given" constraints
* Close it by wrapping in a forall over all its free variables

Tricky stuff, but all the more important to write down.  Does anyone have a better way to characterise it?

Reply:
I think "result type" only applies for flat pattern matches. Whenever there is a nested pattern-match, particularly one with existentials or constraints like `MkX`, it's a bit like a new local function that does the inner pattern-match. I could rewrite your `X` example to:

```hs
data X where { MkX :: Eq x => T x ->X }   -- x is existential

f :: X -> blah
f (MkX tx) = g tx
  where
    g :: Eq x => T x -> blah
    g (K1 x) = ...
    g (K2 p q) = ...
```

Yes, we simply accumulate Givens. Let's see what happens for full-blown pat syns:

```hs
-- required constraints T, provided constraints P
pattern Ps :: forall r. R r => forall p. P p => T r p -> X r
pattern Ps = ... -- unimportant

f :: forall r. R r => X r -> blah
f (Ps trp) = g trp
  where
    g :: forall r p. (R r, P p) => T r p -> blah
    g (K1 x) = ...
    g (K2 p q) = ...
```

So I suppose we may even regard required constraints simply as Givens.

(Things are different for deciding which ConLikes are applicable at a particular result type, though, but that is not of concern to the semantics of the result type signature of a COMPLETE set.)